packages/ckeditor5-list/src/list/utils/listwalker.ts

```diff
@@ -191,47 +191,69 @@ export default class ListWalker {
 
 /**
  * Iterates sibling list blocks starting from the given node.
- *
- * @internal
- * @param node The model node.
- * @param direction Iteration direction.
- * @returns The object with `node` and `previous` {@link module:engine/model/element~Element blocks}.
  */
-export function* iterateSiblingListBlocks(
-	node: Node | null,
-	direction: 'forward' | 'backward' = 'forward'
-): IterableIterator<ListIteratorValue> {
-	const isForward = direction == 'forward';
-	const previousNodesByIndent: Array<ListElement> = []; // Last seen nodes of lower indented lists.
-	let previous = null;
+export class SiblingListBlocksIterator implements IterableIterator<ListIteratorValue> {
+	private _node: Node | null;
+	private _isForward: boolean;
+	private _previousNodesByIndent: Array<ListElement> = [];
+	private _previous: ListElement | null = null;
+	private _previousNodeIndent: number | null = null;
+
+	/**
+	 * @param node The model node.
+	 * @param direction Iteration direction.
+	 */
+	constructor(
+		node: Node | null,
+		direction: 'forward' | 'backward' = 'forward'
+	) {
+		this._node = node;
+		this._isForward = direction === 'forward';
+	}
 
-	while ( isListItemBlock( node ) ) {
-		let previousNodeInList = null; // It's like `previous` but has the same indent as current node.
+	public [ Symbol.iterator ](): IterableIterator<ListIteratorValue> {
+		return this;
+	}
 
-		if ( previous ) {
-			const nodeIndent = node.getAttribute( 'listIndent' );
-			const previousNodeIndent = previous.getAttribute( 'listIndent' );
+	public next(): IteratorResult<ListIteratorValue> {
+		if ( !isListItemBlock( this._node ) ) {
+			return { done: true, value: undefined };
+		}
+
+		let nodeIndent = null;
+		let previousNodeInList: ListElement | null = null;
+
+		if ( this._previous ) {
+			nodeIndent = this._node.getAttribute( 'listIndent' );
+			const previousNodeIndent = this._previousNodeIndent!;
 
 			// Let's find previous node for the same indent.
 			// We're going to need that when we get back to previous indent.
 			if ( nodeIndent > previousNodeIndent ) {
-				previousNodesByIndent[ previousNodeIndent ] = previous;
+				this._previousNodesByIndent[ previousNodeIndent ] = this._previous;
 			}
 			// Restore the one for given indent.
 			else if ( nodeIndent < previousNodeIndent ) {
-				previousNodeInList = previousNodesByIndent[ nodeIndent ];
-				previousNodesByIndent.length = nodeIndent;
+				previousNodeInList = this._previousNodesByIndent[ nodeIndent ];
+				this._previousNodesByIndent.length = nodeIndent;
 			}
 			// Same indent.
 			else {
-				previousNodeInList = previous;
+				previousNodeInList = this._previous;
 			}
 		}
 
-		yield { node, previous, previousNodeInList };
+		const value = {
+			node: this._node,
+			previous: this._previous,
+			previousNodeInList
+		};
+
+		this._previous = this._node as ListElement;
+		this._previousNodeIndent = nodeIndent;
```

Comment:
It's not easy to compare these changes so forgive me if I am making a mistake somewhere here, but.

After first iteration `nodeIndent` is `null`, so `_previousNodeIndent` will be also `null`. In second iteration, in line 228, `previousNodeIndent` will be set to `null`. I think this is different than before.

Reply:
Other than it looks to me that the iteration logic is the same.

Reply:
You're right, fixed.